### DISCUSSION TITLE
Support datetime strings in parse_date

### DISCUSSION THREAD

**Discussion by dhimmel** [pydantic/pydantic#3892] [Feature Request]
https://github.com/samuelcolvin/pydantic/pull/298 added support for parsing a datetime as a date. However, parsing a datetime string as a date is note supported. Both snippets below give a DateError for `invalid date format`:

```py
from datetime import date
from pydantic import BaseModel

# snippet 1
class Model(BaseModel):
    date_field: date
Model(date_field="2004-05-12T00:00:00Z")

# snippet 2
from pydantic.datetime_parse import parse_date
parse_date("2004-05-12T00:00:00Z")
```

I'd like to convert the datetime string to a date, because for this field the HH:MM:SS is always 00:00:00. It seems common for many databases to store fields that should be dates as datetimes. Therefore, it's a common data cleaning step to convert these values back to dates, ideally avoiding any timezone challenges that come from actually parsing the time.

My proposal is for parse_date to recognize strings that start with `YYYY-MM-DD` and use that to create the date.

**Comment by dhimmel** [comment]
In the meantime, I'm using a validator like:

```py
class Model(BaseModel):
    date_field: date

    @validator("date_field", pre=True)
    def timestamp_to_date(cls, v: Any) -> Any:
        """
        Extract the date from a string like '2004-01-01T00:00:00Z'.
        """
        if not v:
            return None
        if not isinstance(v, str):
            raise TypeError(
                f"timestamp_to_date expected a string value, received {v!r}"
            )
        date_str, time_str = v.split("T")
        return date.fromisoformat(date_str)
```

**Comment by 07pepa** [comment]
i would apreciate similar feature (where i could set custom datetime format of timestamp (even my is iso-format but i only need precision by miunute so i do not need parsing of seconds and minutes)

**Comment by dhimmel** [comment]
@samuelcolvin would it be better to create an issue for this feature request? I don't remember why I initially posted this as a discussion, but just encountered this use case anew.

**Reply by dhimmel** [reply]
Noting that https://github.com/pydantic/pydantic/pull/4516 made major changes that might affect date parsing, and is planned to be released as part of [version 2](https://github.com/orgs/pydantic/projects/1). So trying on the current latest main commit:

```
pip install git+https://github.com/pydantic/pydantic@6267ae3538cf7237ff6769174b4c340415e0f444
```

Which produces the following error:

```python-traceback
>>> Model(date_field="2004-05-12T00:00:00Z")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dhimmel-rs/miniforge3/lib/python3.10/site-packages/pydantic/main.py", line 161, in __init__
    values, fields_set = __pydantic_self__.__pydantic_validator__.validate_python(data)
pydantic_core._pydantic_core.ValidationError: 1 validation error for Model
date_field
  Datetimes provided to dates should have zero time - e.g. be exact dates [type=date_from_datetime_inexact, input_value='2004-05-12T00:00:00Z', input_type=str]
```

So my updated feature request for pydantic v2 would be either an option to allow datetime to date conversion by dropping time or perhaps to not raise `date_from_datetime_inexact` when `T00:00:00` is observed.

Also noting that with v2, the datetime object to date conversion will stop working also due to `date_from_datetime_inexact`:

```python
Model(date_field=datetime.datetime.now())
```

Noting the underlying [rust code](https://github.com/pydantic/pydantic-core/blob/8d0678cd0cc1e68cafbf104f7a3385467946d7f1/src/validators/date.rs#L144-L148) that raises `ErrorType::DateFromDatetimeInexact`. Here we see that timezoneless datetimes with `T00:00:00` are accepted, so the following will parse without error:

```py
Model(date_field="2004-05-12T00:00:00")
```

**Reply by dhimmel** [reply]
Some more references on v2:

- https://docs.pydantic.dev/blog/pydantic-v2/
- https://docs.rs/speedate/latest/speedate/

**Reply by dmontagu** [reply]
@dhimmel I think this is a reasonable request, though I will say I also see the perspective that including a timezone (i.e., adding the Z on the end) feels to me like an indication that the timestamp is timezone aware, and therefore the date may depend on a timezone.

My first reaction is that I'm nervous to give special privilege to the "Z" timezone and allow it to pass validation when populating a date, but disallow it for other timezones. (Even if this feels like the "best" timezone to ignore, I'm not sure that ignoring any timezone is the best idea.)

My suggestion would be that, if you want to allow `2004-05-12T00:00:00Z` but disallow `2004-05-12T00:00:00+01:00`, I would maybe just add a pre-validator that strips off the `Z` if it's trailing. I think you should be able to subclass `datetime` with something that adds custom validators, or possibly even leverage just annotations, to create a reusable type annotation that would do this Z-stripping.

I'm open to different opinions though if you think it is really important to have "official" support for parsing a `...00:00:00Z` datetime as a date. In that case, I would suggest creating an issue for easier tracking.

**Reply by dhimmel** [reply]
Hey @dmontagu thanks for the input. I just recently learned about pydantic v2, so let's update my example to assume pydantic 2 and set `strict=False` (which I think is the default):

```py
from datetime import date
from pydantic import BaseModel, Field

class Model(BaseModel):
    date_field: date = Field(strict=False)

# this works
Model(date_field="2004-05-12T00:00:00")

# this following fail due to date_from_datetime_inexact
Model(date_field="2004-05-12T00:00:00Z")
Model(date_field="2004-05-12T00:00:00+01:00")
Model(date_field="2004-05-12T11:11:11")
```

I think it might be nice to have an even less strict mode that would succeed on all these datetime like strings to return the date 2004-05-12. Although I can see how the current strict=False behavior is comparable to converting a float with a fractional part to an int, which will fail now due to `int_from_float` / "Input should be a valid integer, got a number with a fractional part".

So maybe what I am suggesting is 3 levels of strictness:

1. super strict like `strict=True` where types must match
2. less strict like `strict=False` where coercion is allowed as long as there is no data loss
3. permissive where potential data loss is acceptable as long as there is a route from the input value to the output type

Is there anything like option 3 in pydantic 2?

**Reply by dmontagu** [reply]
I'm not confident on all the details of how strictness will behave in pydantic 2, though I think we will likely ensure there is an easy way to do something backwards-compatible. That said, I think you should definitely be able to create a reusable annotation in just a few lines of code that will achieve the coercion logic you prefer on a per-type basis. So, while more "permissive" coercion might require a little more code than it did in v1, it should still be possible.

As we near the v2 release we'll be paying close attention to concerns like this, and will do our best to make sure there's a satisfactory path forward for library users.
